### DISCUSSION TITLE
Emoji/full unicode support in PDF exports

### DISCUSSION THREAD

**Discussion by thegcat** [pretix/pretix#3333] [Ideas]
I guess this is more an ask to switch to (probably) `fpdf2`, as `pypdf` seems to be abandonware, but I wasn't sure how you'd like to have it in the bug tracker.

We regularly have users that use emoji as part of their names and other emoji-things that end up on badges, it would be great if Pretix could support this.

**Comment by raphaelm** [comment]
I'm all for emoji support on badges.

But your guess is quite wrong ;)

1. I don't get why pypdf is abandonware. It's receiving regular releases and every bug we recently found was fixed within days. https://pypi.org/project/pypdf/#history

2. pypdf is not involved in generating our badge PDFs, other than merging in the background image on systems without pdftk or merging multiple documents into one. PDF generation uses reportlab.

reportlab is also not abandoned and is also one of the most mature PDF generation toolkits I have seen.

reportlab apparently added some kind of emoji support back in 2014:
https://docs.reportlab.com/releases/notes/whats-new-31/

It might just be a font issue?

Any changes to badge/ticket generation require great caution though, since we maintain *three* rendering engines (one in JS based on fabric.js for the editor, one in python based on reportlab, and one in Java based on OpenPDF for printing on Android) which should always produce the same results if possible.

**Reply by thegcat** [reply]
> But your guess is quite wrong ;)
> 
> 1. I don't get why pypdf is abandonware. It's receiving regular releases and every bug we recently found was fixed within days. [pypi.org/project/pypdf/#history](https://pypi.org/project/pypdf/#history)

Heh, I followed a rabbit hole and landed in a pypdf issue regarding emoji that stated it was more or less abandoned. It seems I should have verified that :-)

> 2. pypdf is not involved in generating our badge PDFs, other than merging in the background image on systems without pdftk or merging multiple documents into one. PDF generation uses reportlab.

Ah.

> It might just be a font issue?

That was my first thought and tried out a few different fronts from the freefont-pack, I'd have expected Ubuntu, Fira, Noto or Roboto to have them, but they all printed boxes in the PDF.

> Any changes to badge/ticket generation require great caution though, since we maintain _three_ rendering engines (one in JS based on fabric.js for the editor, one in python based on reportlab, and one in Java based on OpenPDF for printing on Android) which should always produce the same results if possible.

Yes, I expected this to be a rather core/important part, which is why I started with a discussion instead of a bug report.
